Fix several warnings

Martchus opened the pull request:
* Avoid implicit signedness conversions where possible
* Avoid c-style casts
* Avoid warnings about bypassing virtual dispatch in destructors

Martchus committed:
Fix several warnings

* Avoid implicit signedness conversions where possible
* Avoid c-style casts

Martchus committed:
Avoid warnings about bypassing virtual dispatch in destructors

aminya commented:
@ahupowerdns Could you take a look at this?

aminya commented on lmdb-typed.hh:
```diff
@@ -153,15 +153,26 @@ struct index_on_function : LMDBIndexOps<Class, Type, index_on_function<Class, Ty
   template<typename Class>
-  void put(MDBRWTransaction& txn, const Class& t, uint32_t id, int flags=0)
-  {}
+  void put(MDBRWTransaction& txn, const Class& t, uint32_t id, unsigned int flags=0)
+  {
+      (void)txn;
+      (void)t;
+      (void)id;
+      (void)flags;
+  }
```
```suggestion
  void put([[maybe_unused]] MDBRWTransaction& txn, [[maybe_unused]] const Class& t, [[maybe_unused]] uint32_t id, [[maybe_unused]] unsigned int flags=0)
  {}
```

aminya reviewed (commented):
Instead of casting to void it is better to use `[[maybe_unused]]`

Martchus commented:
I suppose when making that PR I didn't want to raise the C++ standard to 17 but yes, `[[maybe_unused]]` would be generally preferable.

Note that this repository seems generally unmaintained so I wouldn't get my hopes up that this change is merged anytime soon. I personally just created a "custom" branch on my own fork (see https://github.com/Martchus/lmdb-safe). It is already 66 commits ahead of this one. I picked up some commits from pending PRs and did many adjustments myself. You might not want all changes from my branch but be sure to have a look.

aminya commented:
Thanks! Maybe @ahupowerdns can add you as a maintainer so you can take over the project. It would be great to keep this effort going. Alternatively, we can create a new organization and put the code there.

Martchus commented:
Maybe. Note that I've introduced a dependency to https://github.com/Martchus/cpp-utilities because I didn't want to C&P lots of build system code. That could be problematic for some users (although the library is *not* a runtime dependency, it is just used for CMake modules and a few headers). I've also introduced usage of https://github.com/Martchus/reflective-rapidjson but that's only optional (you can still use any serialization you want). Some of my changes might also be API breaks.
